internal/management/controller/database_controller.go

```diff
@@ -0,0 +1,393 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	apiv1 "github.com/cloudnative-pg/cloudnative-pg/api/v1"
+	"github.com/cloudnative-pg/cloudnative-pg/pkg/management/log"
+	"github.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres"
+	"github.com/cloudnative-pg/cloudnative-pg/pkg/resources/instance"
+	"github.com/cloudnative-pg/cloudnative-pg/pkg/utils"
+)
+
+// DatabaseReconciler reconciles a Database object
+type DatabaseReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	instance *postgres.Instance
+}
+
+// errClusterIsReplica is raised when the database object
+// cannot be reconciled because it belongs to a replica cluster
+var errClusterIsReplica = fmt.Errorf("waiting for the cluster to become primary")
+
+// databaseReconciliationInterval is the time between the
+// database reconciliation loop failures
+const databaseReconciliationInterval = 30 * time.Second
+
+// databaseFinalizerName is the name of the finalizer
+// triggering the deletion of the database
+const databaseFinalizerName = utils.MetadataNamespace + "/deleteDatabase"
+
+// +kubebuilder:rbac:groups=postgresql.cnpg.io,resources=databases,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=postgresql.cnpg.io,resources=databases/status,verbs=get;update;patch
+
+// Reconcile is the database reconciliation loop
+func (r *DatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	contextLogger, ctx := log.SetupLogger(ctx)
+
+	contextLogger.Debug("Reconciliation loop start")
+	defer func() {
+		contextLogger.Debug("Reconciliation loop end")
+	}()
+
+	// Get the database object
+	var database apiv1.Database
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      req.Name,
+	}, &database); err != nil {
+		// This is a deleted object, there's nothing
+		// to do since we don't manage any finalizers.
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// This is not for me!
+	if database.Spec.ClusterRef.Name != r.instance.ClusterName {
+		return ctrl.Result{}, nil
+	}
+
+	// If everything is reconciled, we're done here
+	if database.Generation == database.Status.ObservedGeneration {
+		return ctrl.Result{}, nil
+	}
```

Comment:
What if there's drift on the side of the database because a user made changes there?
If I'm understanding right, as long as the CR didn't change, this would be left alone.